src/lib/formats/fs_hp98x5.cpp

```diff
@@ -0,0 +1,1547 @@
+{
+	*it++ = u8(w >> 8);
+	*it = u8(w & 0xff);
+}
+
+std::vector<u8> hp98x5_impl::get_sector_range(lba_t first, unsigned size) const
+{
+	std::vector<u8> res;
+
+	res.resize(size * SECTOR_SIZE);
+
+	u8 *ptr = res.data();
+
+	for (lba_t idx = first; idx < first + size; idx++) {
+		auto data_sect = m_blockdev.get(u32(idx));
+		memcpy(ptr, data_sect.rodata(), SECTOR_SIZE);
+		ptr += SECTOR_SIZE;
+	}
+	return res;
+}
+
+void hp98x5_impl::store_sector_range(lba_t first, const std::vector<u8>& data)
+{
+	const u8 *ptr = data.data();
+	u32 to_go = u32(data.size());
+	unsigned sects = (data.size() + SECTOR_SIZE - 1) / SECTOR_SIZE;
+	for (lba_t idx = first; idx < first + sects; idx++) {
+		u32 count = std::min<u32>(to_go, SECTOR_SIZE);
+		auto blk = m_blockdev.get(u32(idx));
+		blk.copy(0, ptr, count);
+		ptr += count;
+		to_go -= count;
+	}
+}
+
+void hp98x5_impl::check_av_data(const std::vector<u8>& in, const std::vector<u8>::const_iterator& it, int min_av)
+{
+	if (in.cend() - it < min_av) {
+		throw std::runtime_error("DATA file too short");
+	}
+}
+
+u16 hp98x5_impl::get_u16(const std::vector<u8>& in, std::vector<u8>::const_iterator& it)
+{
+	check_av_data(in, it, 2);
+	auto tmp = r16b(it);
+	it += 2;
+	return tmp;
+}
+
+std::vector<u8> hp98x5_impl::convert_data_2_txt(const std::vector<u8>& in, unsigned bpr)
+{
+	auto it = in.cbegin();
+	std::vector<u8> accum;
+
+	u16 rec_type;
+	unsigned accum_len = 0;
+
+	do {
+		// Get record type
+		rec_type = get_u16(in, it);
+		switch (rec_type) {
+		case REC_TYPE_EOR:
+			// End of record: just skip it
+			break;
+
+		case REC_TYPE_FULLSTR:
+			// A string in a single piece
+		case REC_TYPE_1STSTR:
+			// First piece of a split string
+		case REC_TYPE_MIDSTR:
+			// Mid piece(s) of a split string
+		case REC_TYPE_ENDSTR: {
+			// Closing piece of a split string
+			if (((rec_type == REC_TYPE_FULLSTR || rec_type == REC_TYPE_1STSTR) && accum_len > 0) ||
+				((rec_type == REC_TYPE_MIDSTR || rec_type == REC_TYPE_ENDSTR) && accum_len == 0)) {
+				throw std::runtime_error("Wrong sequence of string pieces");
+			}
+
+			unsigned tmp_len = (unsigned)get_u16(in, it);
+
+			if (rec_type == REC_TYPE_FULLSTR || rec_type == REC_TYPE_1STSTR) {
+				accum_len = tmp_len;
+			} else if (tmp_len != accum_len) {
+				throw std::runtime_error("Wrong length of string piece");
+			}
+
+			unsigned len_to_eor = bpr - (it - in.cbegin()) % bpr;
+			unsigned rec_len;
+			if (rec_type == REC_TYPE_FULLSTR || rec_type == REC_TYPE_ENDSTR) {
+				if (accum_len > len_to_eor) {
+					throw std::runtime_error("Wrong length of string piece");
+				}
+				rec_len = accum_len;
+			} else {
+				if (accum_len <= len_to_eor) {
+					throw std::runtime_error("Wrong length of string piece");
+				}
+				rec_len = len_to_eor;
+			}
+			check_av_data(in, it, rec_len);
+			// Sanitize and copy string
+			for (unsigned i = 0; i < rec_len; i++) {
+				u8 c = *it++;
+				if (!isascii(c) || !isprint(c)) {
```

Comment:
You shouldn’t use `isascii` as it’s deprecated in POSIX and has weird semantics.  You shouldn’t rely on `isprint` because it’s locale-dependent.  Also, MAME uses UTF-8 for text in general.  You’ll need your own character handling helpers that make a best guess at the encoding of the data.